### DISCUSSION TITLE
Excessive memory usage among HA servers during member restart

### DISCUSSION THREAD

**Discussion by kskalski** [k3s-io/k3s#7337] [Q&A] [answered]
**Environmental Info:**
K3s Version: `v1.27.1-rc1+k3s1` (but also observed in earlier versions, e.g. `v1.25.9+k3s1`)

Node(s) CPU architecture, OS, and Version: 
`Linux vps-96dabacf 5.15.0-70-generic #77-Ubuntu SMP Tue Mar 21 14:02:37 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux`

Cluster Configuration:
3 servers running in HA mode, all with config
```
ExecStart=/usr/local/bin/k3s \
    server \
	'--disable=traefik' \
	'--disable=servicelb' \
```
two of them have 2 cores + 4GB ram, 1 has 4 cores and 8GB ram, they run on VPSed in the same cluster with internal IPs in range `10.42.*.*` and public IPs in range `146.*.*.*`, public IPs are used in etcd membership list.

**Describe the bug:**
Normally k3s in HA server mode in my setup uses around 700-800MB of ram, however recently I observed the memory usage growth to 3000MB or more, which so far seems to be related to *other* servers in the membership list being restarted.
This took me by surprise: the memory usage didn't fit with my other payload -> servers started swapping activity -> cpu saturated -> etcd/raft messages started getting high latency or drops -> lots of error messages flooded logs and eventually the server gets killed - maybe by oom killer or by accumulating some errors and deciding it's enough.

Note the proper provisioning of resources to servers require me to keep all the slack memory free, so effectively it requires counting k3s server reserved memory at ~3GB plus some margin.

**Steps To Reproduce:**
My installation is a pre-existing set-up which I upgraded from earlier versions, and previously I disabled flannel and used calico with wireguard, only recently re-configured cluster to use default flannel backend and disabled calico (not sure this makes a difference, but possibly there is some crust accumulated in my cluster that made it behave this way, though now it happens on all the nodes).

* install 3 k3s servers pointing the subsequent servers to the first one using  `--server=https://public_ip:6443` flag
* restart one of servers
* observe other servers memory usage grow from ~800MB to ~2000-3000MB, also cpu usage spikes to 100% on all cores, but if machine has enough ram it goes down after a few seconds (high RES memory stays though)

![image](https://user-images.githubusercontent.com/775012/233658554-1b4d8b0f-a162-46e2-81db-5c32b4b27775.png)

vs
most recently restarted server:
![image](https://user-images.githubusercontent.com/775012/233658718-ee638e25-3a3b-4b87-b185-ea0c233db6b1.png)

**Expected behavior:**
Ideally restarting servers and running them in any situation (stable or degraded health of other members) should keep memory usage at some reasonable level. Clearly, establishing new leader requires additional operations outside of regular function in stable mode, but I think it shouldn't be more than tens of MB, maybe 100-200 MB.

**Actual behavior:**
It seems that in stable situation the most recent restarted / re-joining server grows to its regular 800MB and when no more restarts / communication issues arise, it stays this way. However in the meantime the other 2 servers whether they used <1GB or more, they start rapid memory usage expansion, usually to 2300MB, sometimes more. After situation stabilizes, they usually stay at that elevated memory usage.

**Additional context / logs:**
I'm attaching log from the server experiencing memory growth (the healthy one that observed its peer being restarted)
file:///home/nazgul/rep/gradcap/backtest-py/log.txt

**Comment by brandond** [comment, marked as answer]
Check out https://etcd.io/docs/v3.5/tuning/ - in particular the sections on the snapshot count. However, since you're seeing the memory utilization only go up when one of the nodes is offline, I suspect that you might be running into issues with the peer's buffer growing, and golang not freeing the memory back to the host operating system once the peer comes back online. I'm not aware of any way to tune the etcd buffer sizes.

You might try:
* Disabling swap on your nodes. Best practice for Kubernetes is to disable swap, especially since you mention that high swap IO is what eventually kills your node.
* Setting a memory limit on the k3s systemd service. Without a memory limit, golang will not aggressively garbage collect and free memory back to the host OS, which will make your swap problems worse.

**Reply by kskalski** [reply]
Thanks, setting 
```
MemoryHigh=1000M
MemoryMax=1G
```
in `/etc/systemd/system/k3s.service` indeed allows to cap the memory usage expansion.
I'm still seeing a lot of cpu usage now, maybe attributing it to swap was a red-herring (I don't actually have swap partitions, but `top` showed kswapd with high cpu usage, maybe there is a way to disable it completely). Currently there is constant high cpu usage and top is showing it for `containerd-shim` processes (a bit confused if this just means some of my payload containers or something in the infrastructure)

![image](https://user-images.githubusercontent.com/775012/233694016-0b1f89a7-daff-4e0b-9cad-55be2c125ae1.png)

**Reply by brandond** [reply]
Be aware that setting the memory limit too low will force go to run garbage collection more frequently, which will push up CPU utilization and reduce overall system efficiency. Don't be too aggressive with capping that, as it directly impacts datastore and apiserver performance.

**Reply by kskalski** [reply]
Yeah, clearly capping the memory pushes the system to burn more CPU. 

Actually my problem is that the cluster easily enters an unstable state, possibly with some feedback loops making it unable to recover, where
* 100% of cpu is burned on all servers (by k3s process plus containerd-shim) and 
* often all available memory is grabbed by k3s process (either within caps or all available memory on server when uncapped).

Even with no caps and ~3GB free memory on each server sometimes I'm observing one or all servers unable to recover and constantly burning 100% cpu... At least coordinated reboot of all the machines at the same time seems to restart the system in such a way, that k3s is doing a minute of high cpu burn, but recovers then.

I'm not completely sure what are the feedback loops, but the fact that many of the pods enter crash loop probably increases pressure on etcd and prevents recovery.
Going forward I would like to try:
* figure out if `containerd-shim` processes burning lots CPU are attributable to some of my in-cluster payloads/pods (longhorn is a suspect here) - in the process tree I can't see what they are connected to, but they have the `--namespace k3s.io --id 334534....` in command line
* experiment a bit more with relationship of capped k3s memory and how easy it is for system to melt down
* see if tuning snapshot count makes a difference

**Reply by brandond** [reply]
Longhorn is pretty IO, thread, and memory intensive. You might be pushing it trying to run k3s + etcd + LH on just 2 cores and 4GB of RAM. The fact that you have a system load of 71 suggests that there are more processes vying for time than you can successfully schedule onto just 2 cores.

**Reply by kskalski** [reply]
I suspect this is a disk contention issue as you suggested also [here](https://github.com/k3s-io/k3s/issues/7343#issuecomment-1520643347), since a bigger node (4 cores and 8GB of ram) sometimes end up in the same failure mode, which I also observed to be IO bound:
- `iotop` shows >600MB/s read rate, close to half of it coming from k3s server (etcd?) and the rest from containers / longhorn
- there is almost no `app` cpu usage, all cores are burning `sys` cpu, which I suppose happens when VPS server does a lot of IO
- etcd and longhorn run on the same disk, which apparently makes them enter some insane read/retry loop without any back-off or throttling

I also tried increasing etcd args to `heatbeat-interval=500,election-timeout=5000`, but didn't observe any improvement in recovery. What makes recovery more likely (though not in 100% of cases) is not running longhorn components or rebooting all the machines at the same time.

For the record, my `/var/lib/rancher/k3s/server/db/etcd/member/snap/db` file is ~150MB, and the snapshots in that dir are around 400KB and created every ~25min. Is leader `etcd` transferring the whole `db` file to a joining member? It isn't really too big, but with tight enough timeout maybe it's failing to read it quickly and then retries hammering the disk?
